jigx-com/fastlane-jigx
# Match flexibility improvements

### Checklist

### Motivation and Context


The main issue is that Match doesn't work that good outside of the "golden-path". Someone might have their account already set up with some existing apps or may already have a 3rd party involved also using fastlane for other apps, etc.. in their account and we cant just nuke it. We also need to support people importing their certs manually regularly without stuff breaking.

This is currently not supported and results in a big mess.

The root cause is that match heavily relies on certain preconditions and does not perform more complex checks like actually checking certificate, provisioning profile contents and matching them according to it.

This leads to problems when people import manually created certificates as all crucial info on which match relies right now such as the Apple Cert ID (contained in the name when created by match using API access, not present at all when created manually) is lost.

### Description


The main changes include using the serial number of certificates/profiles to check for their validity on the dev portal (instead of the file name - apple cert id/provisioning profile name) and matching/selecting provisioning profiles by checking the embedded certificates (instead of relying on a single predetermined workflow and selecting the first one present in storage).

The unit tests were edited to account for the changes. It was also tested live in our dev environment.

### Testing Steps


Since the main goal is to make it more robust in all situations, not being able to "break" it by not following the match docs steps should be counted as a success. 
Concrete test cases include: 
1. not using a clean account
2. not nuking
3. importing manually created certs and mixing it with API access to then automatically setup some more of them
